File: 백준/Silver/11729. 하노이 탑 이동 순서/하노이 탑 이동 순서.py
def hanoi(n, start, aux, goal): # 시작 지점, 보조 지점, 목표지점
  if n == 1:
    print(start, goal)
    return
  hanoi(n - 1, start, goal, aux)
  print(start, goal)
  hanoi(n - 1, aux, start, goal)


n = int(input())

print(2 ** n - 1)
hanoi(n, 1, 2, 3)

File: 백준/Silver/11729. 하노이 탑 이동 순서/하노이 탑 이동 순서.cc
#include <stdio.h>
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include <set>
#include <ostream>
#include <string>
#include <cstring>
#include <sstream>
#include <time.h>
#include <queue>
#include <cstdlib>
#include <stdbool.h>
#include <map>

using namespace std;

int n;
vector<pair<int, int>> vp;

void hanoi(int num, int start, int path, int end){
    if (num == 1){
        vp.push_back(pair<int, int> (start, end));
        return ;
    }
    hanoi(num - 1, start, end, path);
    vp.push_back(pair<int, int> (start, end));
    hanoi(num - 1, path, start, end);
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    
    cin >> n;
    
    hanoi(n, 1, 2, 3);
    cout << vp.size() << endl;
    for (int i = 0; i < vp.size(); i++){
        cout << vp[i].first << " " << vp[i].second << '\n';
    }
    
    return 0;
}
